### ISSUE TITLE
Add hot-keys for more RD functions

### ISSUE THREAD

**Issue by daFreeMan** [rubberduck-vba/Rubberduck#3070]
The current set of hot-keys are handy, but are only available for a limited set of RD functions. It would be great if more were added.

There's no _critical_ need to enable all hot-keys by default, or even make default assignments, though there's no issue if this does happen, so there's no need for the coder to think up all the combinations...

Commands in general could be bound to hot-keys, and, possibly, allow for plugins to be bound to hot-keys. (Hat tip to [@Vogel612](https://chat.stackexchange.com/transcript/message/38132557#38132557))

**Comment by Vogel612** [comment]
Inspiration for this could be taken from how Inspections are discovered

**Comment by Hosch250** [comment]
What would be cool would be great is if we could do it VS style. We expose a list of commands, and allow the user to set any hotkey combination by just pressing the keys, which also overrides the default binding.
![new mockup 1](https://user-images.githubusercontent.com/6299719/27257556-c95c2208-53a0-11e7-8a38-ed18dbf33fde.png)

**Comment by retailcoder** [comment]
Currently, adding a hotkey for a command implies:

 - Adding an enum member in [RubberduckHotkey.cs](https://github.com/rubberduck-vba/Rubberduck/blob/next/RetailCoder.VBE/Settings/RubberduckHotkey.cs). This ensures name consistency in `HotkeySettings`, where the `ToString` representation of these enums is used.
 - Add a description for the hotkey in RubberduckUI.resx. This resource key is used in `HotkeySettings` to get a localized name for the command.
 - Adding a default `HotkeySetting` instance in [HotkeySettings.cs](https://github.com/rubberduck-vba/Rubberduck/blob/next/RetailCoder.VBE/Settings/HotkeySettings.cs#L15-L29). This determines the default hotkey for the command, and makes the setting dialog aware of the hotkey.
 - Adding a `HotkeySetting` instance in [HotkeySettingsTests.cs](https://github.com/rubberduck-vba/Rubberduck/blob/next/RubberduckTests/Settings/HotkeySettingsTests.cs), to provide data for the tests in that class.
 - Adding a `public override RubberduckHotkey Hotkey => RubberduckHotkey.SomeHotkey` member in the command implementation, to return the `RubberduckHotkey` enum for that command. The `RubberduckHooks` class gets all commands constructor-injected, and this property is used to retrieve the command associated with a given `RubberduckHotkey` enum value. There should be a test that fails given two commands returning the same enum value, for the hooks only get whichever first command has a `Hotkey` property that returns the `RubberduckHotkey` enum it's looking for.

Therefore, implementing this would be awesome, but it pretty much entails a complete redesign of how hotkeys work in Rubberduck.

**Comment by daFreeMan** [comment]
I'm not sure I follow why this would take redesigning the hotkey setup. It looks like there's a process in place (even if this is the only place it's documented and even if it's 5 steps), so "all" one has to do is "just" update the 5 locations you identified and that should be it, right? Maybe there's a 6th step in updating the setting dialog, but still, it seems straightforward, and since they work now, I'm not sure I understand why the system would have to be changed to make new ones work.

**Comment by retailcoder** [comment]
@daFreeMan because it's a lot of places to go and modify, only to enable a hotkey for a command. If only a few select commands need hotkeys, it's a fine "process" - but if we want every command to have a hotkey mapping *by default* then we need a better solution, one that takes care of most of the plumbing for us.

The `RubberduckHotkey` enum smells, for one: I shouldn't need to modify an enum every single time I make a new hotkey-enabled command. I'd want `CommandBase` to take care of that, in one place, for *all* commands. Like inspections, perhaps commands need a `Type` property that simply returns a `Type` instance (`GetType` would work, if we weren't using DI interception and injecting proxy types), and we can use that type's `Name` the way we currently use the enum, for the same purposes.

I see this design issue similar to how adding a new inspection was too complicated (ref. #3021) - in fact there are glaring similarities between what `GetDefaultCodeInspections` looked like not too long ago, and what `HotkeySettings` is doing. Perhaps each hotkey-enabled command could have a `public override HotkeySetting DefaultHotkey { get; }` property that returns the default hotkey configuration, that way all `HotkeySettings` would have to do is iterate the `IEnumerable<CommandBase>` and *get* the default hotkey configurations from each individual command.

Basically, as it stands it's too scattered to scale nicely.

**Comment by rkapka** [comment]
Hey! I'm trying to redesign the hotkeys, but I cannot for the life of me figure out how the `IEnumerable<CommandBase>` gets injected into `RubberduckHooks`. Can anyone give me a hint? The constructor is never called directly, I cannot find a call via reflection and I cannot find an appropriate line in the `RubberduckIoCInstaller` class...

By the way, I'm trying to inject the same collection into `HotkeySettings` using property injection. I don't think constructor injection is an option, since `HotkeySettings` is used as `XmlPersistanceService<T>`'s type parameter and `T` has the `new` type constraint. Unless we remove the constraint, which doesn't generate any compiler errors. Is there a specific reason why this constraint has to be there?

**Comment by MDoerner** [comment]
@rkapka The way it is set up CW injects into the `IEnumerable<CommandBase>` parameter everything that got registered to `CommandBase`.

If you want to inject them into a property, you can use `.OnCreate((kernel, item) => item.PropertyName = kernel.Resolve<IEnumerable<CommandBase>>())`.

**Comment by rkapka** [comment]
Closing as discussed in #3534.
